### ISSUE TITLE
Unknown error when invalid $ref is used

### ISSUE THREAD

**Issue by saharj** [apigee-127/sway#84]
_From @ePaul on May 31, 2016 17:1_

**My Swagger**

``` yaml
swagger: '2.0'
info:
  version: 0.0.0
  title: Simple API
paths:
  /:
    get:
      responses:
        200:
          $ref: "#responses/example"
```

(Yes, I know that it is invalid, that's the point.)
- Swagger Editor version(from Help > About menu): 2.10.1 (currently on editor.swagger.io)
- My browser: Chromium 50.0.2661.102 Ubuntu 16.04 (64-bit)

**Issue**

If I have an invalid JSON pointer in one of my `$ref` attribute, the editor shows an non-helpful error message:

```
✖ Unknown Error
ptr must be a JSON Pointer
Jump to line
Details
 Object
message:  "ptr must be a JSON Pointer"
code:  "UNCAUGHT_SWAY_WORKER_ERROR"
level: 900
type:  "Unknown Error"
description:  "ptr must be a JSON Pointer"
```

![swagger-editor-error](https://cloud.githubusercontent.com/assets/645859/15683125/197c2746-2761-11e6-8d18-9cfa16980b92.png)

The "Jump to line" link jumps to some random line (in the example above line 7, which has just the `/`).

**Expected behavior**

It would be nice if that error message showed in which $ref the error occurred (as a line number, and/or as a path to the problematic place, as for other validation errors), and maybe also said in which way the value is not a JSON pointer (in the case here, the `/` is missing after the `#`).

In a real-live API one can have hundreds of `$ref`s, and many API designers even might not know that those are JSON pointers (i.e. that something is wrong with the `$ref`s).

_Copied from original issue: swagger-api/swagger-editor#955_

**Comment by saharj** [comment]
_From @kgalli on May 31, 2016 17:13_

👍  This would be a very nice improvement.

**Comment by whitlockjc** [comment]
Line numbers and column numbers won't happen as that would require generating an AST and getting that information.  The work to do this is very simple, as I've got a patch to do this locally, but it would add even more bulk to an already hefty library.  I'll revisit this when I replace json-schema-faker with something else.

As for the error you're seeing, I've created a json-refs issue to remedy this.

**Comment by saharj** [comment]
The line number and column number are not the issue here. The issue is that Sway throws error instead of giving us the proper error message.

**Comment by whitlockjc** [comment]
I realize that, which is why I said what I did.  You said _"It would be nice if that error message showed in which $ref the error occurred (as a line number, and/or as a path to the problematic place, as for other validation errors)"_ and so I just clearly explained that I can't do that right now.  I also said _"As for the error you're seeing, I've created a json-refs issue to remedy this."_  I'm not sure how this wasn't clear.

**Comment by saharj** [comment]
I'm not sure if you realized it but I moved this issue from Swagger Editor using [Github Issue Mover](https://github-issue-mover.appspot.com). 

Original poster is asking for line numbers etc but the real issue is that Sway was throwing an error instead of reporting an error which I believe you've found the root cause and opened an issue in JSON Refs.

**Comment by ePaul** [comment]
I (as the original poster of https://github.com/swagger-api/swagger-editor/issues/955) was asking for line numbers because other kinds of errors shown in Swagger-Editor get line numbers (and other indications what is going wrong). Sorry, I was not aware that this was an issue caused by some underlying library which might not support this.

**Comment by whitlockjc** [comment]
@ePaul There is no problem with your request for column/line numbers in the errors.  It's a valid request and one that wouldn't be very hard to implement.  As mentioned above, I've implemented it already locally but due to dependencies, it adds a bit of size to an already large library.  Most of that size is due to json-schema-faker and lodash.  The first dependency would require me to use another library.  Fixing lodash's size is easy.  As for the problem with the JSON Pointer error not being verbose enough, this lies in the json-refs project and an appropriate issue has been created: https://github.com/whitlockjc/json-refs/issues/85 Please don't feel like your request is in any way a problem.

**Comment by whitlockjc** [comment]
@saharj My confusion is not with how the request was copied/ported from one repo to this one or with the original request itself or with my understanding of the request.  What I don't understand is why you felt the need to [re-explain what the issue is](https://github.com/apigee-127/sway/issues/84#issuecomment-223702109) **after** I have already clearly responded with an answer to the issue.  _(You even [do it again here](https://github.com/apigee-127/sway/issues/84#issuecomment-223737421).)_  It's comes off as if you're saying I didn't read the issue in the first place or didn't understand what was being requested.  But I did fully read the issue, I did understand it and I responded clearly, including creating [an issue](https://github.com/whitlockjc/json-refs/issues/85) in the appropriate location which shows I understood what was the issue was.

**Comment by petracles** [comment]
Would really appreciate it if this is addressed - debugging my Swagger doc without *any* stacktrace has been a nightmare.
<img width="812" alt="screen shot 2016-07-15 at 5 37 04 pm" src="https://cloud.githubusercontent.com/assets/4413199/16891691/cc70056a-4ab2-11e6-9075-be33c67b62fb.png">


All help appreciated! Thanks :)

**Comment by whitlockjc** [comment]
It's being addressed in the next release but there is a more pressing issue in json-refs that needs to be resolved first.  The plan is to get a json-refs release out, which not only fixes this issue but also fixes the issue(s) I'm referring to, and then we'll update swagger-tools and sway.

**Comment by SwingGuy1024** [comment]
The example above has a missing reference. I'm seeing the same behavior with a malformed reference:  
           `$ref: '#definitions/County'`

After struggling with the error, I finally figured out that I needed to put a slash before the word `definitions`. 
            `$ref: '#/definitions/County'`

I hope your fix covers this case, too.

**Comment by FranckVE** [comment]
I had the same bug and lack of details on what was going wrong. I could only get out of trouble when I made sure that all references to `definitions` were :
like `$ref: "#/definitions/Expenditure"`
instead of ~~`$ref: "#definitions/Expenditure"`~~
that is : adding the slash `/` in the reference to `definitions`: `#/definitions/...`

**Comment by whitlockjc** [comment]
The issue is fixed in `json-refs@v3.0.0` but I've not updated `sway` yet.  I'm working on it.

**Comment by andreamussap** [comment]
I'm using swagger 2, and this happened to me when I forgot to include a slash in the $ref. It sucks that the error msg is not clear, but as soon as I've seen the error I just went back to what I last did and then I found the issue.
(29/12/2017)
![unknow-error-missing-slash](https://user-images.githubusercontent.com/10401080/34438867-aca16782-eca1-11e7-9d41-23653b2c4349.png)

**Comment by whitlockjc** [comment]
https://github.com/whitlockjc/json-refs/issues/130 and https://github.com/whitlockjc/json-refs/issues/131 have been fixed.  I will cut a release and will update `sway` to use it so that the issues reported are resolved.

**Comment by whitlockjc** [comment]
This was fixed in `json-refs@v3.0.4` and commited in https://github.com/apigee-127/sway/commit/cdd8030a7dac5f86de5fb0bdf719b7ed92ada1f0
